README.md
# pix-me

Create PIX codes to use via QRCode or Pix-Copy-Paste

## Installation

```sh
npm i pix-me
```

## Usage

```typescript
import { pix } from "pix-me";

console.log(pix({
  key: "26385709906",
  city: "Cidade",
  name: "FULANO DE TAL",
}));

// 00020126330014br.gov.bcb.pix0111263857099065204000053039865802BR5913FULANO DE TAL6006Cidade62070503***6304151C
```

See tests for more examples

CHANGELOG.md
## [2.0.6](https://github.com/roziscoding/pix-me/compare/v2.0.5...v2.0.6) (2022-09-25)

## [2.0.5](https://github.com/roziscoding/pix-me/compare/v2.0.4...v2.0.5) (2022-09-25)

## [2.0.4](https://github.com/roziscoding/pix-me/compare/v2.0.3...v2.0.4) (2022-09-25)

## [2.0.3](https://github.com/roziscoding/pix-me/compare/v2.0.2...v2.0.3) (2022-09-25)

## [2.0.2](https://github.com/roziscoding/pix-me/compare/v2.0.1...v2.0.2) (2022-09-25)

## [2.0.1](https://github.com/roziscoding/pix-me/compare/v2.0.0...v2.0.1) (2022-09-25)

# [2.0.0](https://github.com/roziscoding/pix-me/compare/v1.1.1...v2.0.0) (2021-09-09)

### Bug Fixes

- **code generation:** fix missing fields
  ([43c74fb](https://github.com/roziscoding/pix-me/commit/43c74fbea7e41cac1d1aa38dca10aceca4c793de))

### Reverts

- reverts changes made by semantic-release
  ([45ac51b](https://github.com/roziscoding/pix-me/commit/45ac51b469c0054b49563e564e517ed8676d6bab))

### BREAKING CHANGES

- **code generation:** The `reusable` param is no longer accepted, since I
  couldn't find a single bank that uses it


test/pix.js
const { expect } = require('chai');
const { pix } = require('../dist');

describe('pix', () => {
  describe('generates valid PIX codes for', () => {
    it('non specified amount', () => {
      expect(
        pix({
          key: '26385709906',
          city: 'Cidade',
          name: 'FULANO DE TAL'
        })
      ).to.equal(
        '00020126330014br.gov.bcb.pix0111263857099065204000053039865802BR5913FULANO DE TAL6006Cidade62070503***6304151C'
      );
    });

    it('email keys', () => {
      expect(
        pix({
          key: 'fulano.de.tal@gmail.com',
          city: 'Cidade',
          name: 'FULANO DE TAL'
        })
      ).to.equal(
        '00020126450014br.gov.bcb.pix0123fulano.de.tal@gmail.com5204000053039865802BR5913FULANO DE TAL6006Cidade62070503***63045E63'
      );
    });

    it('random keys', () => {
      expect(
        pix({
          key: '4b7b4eb3-d426-48f1-8ecf-998bda62c0a1',
          city: 'Cidade',
          name: 'FULANO DE TAL'
        })
      ).to.equal(
        '00020126580014br.gov.bcb.pix01364b7b4eb3-d426-48f1-8ecf-998bda62c0a15204000053039865802BR5913FULANO DE TAL6006Cidade62070503***6304CAC5'
      );
    });

    it('phone keys', () => {
      expect(
        pix({
          key: '+5568375779606',
          city: 'Cidade',
          name: 'FULANO DE TAL'
        })
      ).to.equal(
        '00020126360014br.gov.bcb.pix0114+55683757796065204000053039865802BR5913FULANO DE TAL6006Cidade62070503***630484D0'
      );
    });

    it('cnpj keys', () => {
      expect(
        pix({
          key: '20687705000136',
          city: 'Cidade',
          name: 'FULANO DE TAL'
        })
      ).to.equal(
        '00020126360014br.gov.bcb.pix0114206877050001365204000053039865802BR5913FULANO DE TAL6006Cidade62070503***630472EB'
      );
    });
  });
});


test/pix.test.ts
import { pix } from "../src/pix.ts";

import { assertEquals } from "std/testing/asserts.ts";
import { describe, it } from "std/testing/bdd.ts";

describe("pix", () => {
  describe("generates valid PIX codes for", () => {
    it("non specified amount", () => {
      assertEquals(
        pix({
          key: "26385709906",
          city: "Cidade",
          name: "FULANO DE TAL",
        }),
        "00020126330014br.gov.bcb.pix0111263857099065204000053039865802BR5913FULANO DE TAL6006Cidade62070503***6304151C",
      );
    });

    it("email keys", () => {
      assertEquals(
        pix({
          key: "fulano.de.tal@gmail.com",
          city: "Cidade",
          name: "FULANO DE TAL",
        }),
        "00020126450014br.gov.bcb.pix0123fulano.de.tal@gmail.com5204000053039865802BR5913FULANO DE TAL6006Cidade62070503***63045E63",
      );
    });

    it("random keys", () => {
      assertEquals(
        pix({
          key: "4b7b4eb3-d426-48f1-8ecf-998bda62c0a1",
          city: "Cidade",
          name: "FULANO DE TAL",
        }),
        "00020126580014br.gov.bcb.pix01364b7b4eb3-d426-48f1-8ecf-998bda62c0a15204000053039865802BR5913FULANO DE TAL6006Cidade62070503***6304CAC5",
      );
    });

    it("phone keys", () => {
      assertEquals(
        pix({
          key: "+5568375779606",
          city: "Cidade",
          name: "FULANO DE TAL",
        }),
        "00020126360014br.gov.bcb.pix0114+55683757796065204000053039865802BR5913FULANO DE TAL6006Cidade62070503***630484D0",
      );
    });

    it("cnpj keys", () => {
      ``;
      assertEquals(
        pix({
          key: "20687705000136",
          city: "Cidade",
          name: "FULANO DE TAL",
        }),
        "00020126360014br.gov.bcb.pix0114206877050001365204000053039865802BR5913FULANO DE TAL6006Cidade62070503***630472EB",
      );
    });
  });
});


src/crc16.ts
// Extracted from https://github.com/NascentSecureTech/pix-qrcode-utils/blob/master/packages/emv-merchant-qrcode/src/crc.ts

/**
 * Calculates the CRC for a string
 * @param str String to calculate the CRC for
 * @param invert If the CRC should be inverted
 * @returns The CRC for the string
 */
export function computeCRC(str: string, invert: boolean = false): string {
  const bytes = new TextEncoder().encode(str);

  const crcTable = [
    0x0000,
    0x1021,
    0x2042,
    0x3063,
    0x4084,
    0x50a5,
    0x60c6,
    0x70e7,
    0x8108,
    0x9129,
    0xa14a,
    0xb16b,
    0xc18c,
    0xd1ad,
    0xe1ce,
    0xf1ef,
    0x1231,
    0x0210,
    0x3273,
    0x2252,
    0x52b5,
    0x4294,
    0x72f7,
    0x62d6,
    0x9339,
    0x8318,
    0xb37b,
    0xa35a,
    0xd3bd,
    0xc39c,
    0xf3ff,
    0xe3de,
    0x2462,
    0x3443,
    0x0420,
    0x1401,
    0x64e6,
    0x74c7,
    0x44a4,
    0x5485,
    0xa56a,
    0xb54b,
    0x8528,
    0x9509,
    0xe5ee,
    0xf5cf,
    0xc5ac,
    0xd58d,
    0x3653,
    0x2672,
    0x1611,
    0x0630,
    0x76d7,
    0x66f6,
    0x5695,
    0x46b4,
    0xb75b,
    0xa77a,
    0x9719,
    0x8738,
    0xf7df,
    0xe7fe,
    0xd79d,
    0xc7bc,
    0x48c4,
    0x58e5,
    0x6886,
    0x78a7,
    0x0840,
    0x1861,
    0x2802,
    0x3823,
    0xc9cc,
    0xd9ed,
    0xe98e,
    0xf9af,
    0x8948,
    0x9969,
    0xa90a,
    0xb92b,
    0x5af5,
    0x4ad4,
    0x7ab7,
    0x6a96,
    0x1a71,
    0x0a50,
    0x3a33,
    0x2a12,
    0xdbfd,
    0xcbdc,
    0xfbbf,
    0xeb9e,
    0x9b79,
    0x8b58,
    0xbb3b,
    0xab1a,
    0x6ca6,
    0x7c87,
    0x4ce4,
    0x5cc5,
    0x2c22,
    0x3c03,
    0x0c60,
    0x1c41,
    0xedae,
    0xfd8f,
    0xcdec,
    0xddcd,
    0xad2a,
    0xbd0b,
    0x8d68,
    0x9d49,
    0x7e97,
    0x6eb6,
    0x5ed5,
    0x4ef4,
    0x3e13,
    0x2e32,
    0x1e51,
    0x0e70,
    0xff9f,
    0xefbe,
    0xdfdd,
    0xcffc,
    0xbf1b,
    0xaf3a,
    0x9f59,
    0x8f78,
    0x9188,
    0x81a9,
    0xb1ca,
    0xa1eb,
    0xd10c,
    0xc12d,
    0xf14e,
    0xe16f,
    0x1080,
    0x00a1,
    0x30c2,
    0x20e3,
    0x5004,
    0x4025,
    0x7046,
    0x6067,
    0x83b9,
    0x9398,
    0xa3fb,
    0xb3da,
    0xc33d,
    0xd31c,
    0xe37f,
    0xf35e,
    0x02b1,
    0x1290,
    0x22f3,
    0x32d2,
    0x4235,
    0x5214,
    0x6277,
    0x7256,
    0xb5ea,
    0xa5cb,
    0x95a8,
    0x8589,
    0xf56e,
    0xe54f,
    0xd52c,
    0xc50d,
    0x34e2,
    0x24c3,
    0x14a0,
    0x0481,
    0x7466,
    0x6447,
    0x5424,
    0x4405,
    0xa7db,
    0xb7fa,
    0x8799,
    0x97b8,
    0xe75f,
    0xf77e,
    0xc71d,
    0xd73c,
    0x26d3,
    0x36f2,
    0x0691,
    0x16b0,
    0x6657,
    0x7676,
    0x4615,
    0x5634,
    0xd94c,
    0xc96d,
    0xf90e,
    0xe92f,
    0x99c8,
    0x89e9,
    0xb98a,
    0xa9ab,
    0x5844,
    0x4865,
    0x7806,
    0x6827,
    0x18c0,
    0x08e1,
    0x3882,
    0x28a3,
    0xcb7d,
    0xdb5c,
    0xeb3f,
    0xfb1e,
    0x8bf9,
    0x9bd8,
    0xabbb,
    0xbb9a,
    0x4a75,
    0x5a54,
    0x6a37,
    0x7a16,
    0x0af1,
    0x1ad0,
    0x2ab3,
    0x3a92,
    0xfd2e,
    0xed0f,
    0xdd6c,
    0xcd4d,
    0xbdaa,
    0xad8b,
    0x9de8,
    0x8dc9,
    0x7c26,
    0x6c07,
    0x5c64,
    0x4c45,
    0x3ca2,
    0x2c83,
    0x1ce0,
    0x0cc1,
    0xef1f,
    0xff3e,
    0xcf5d,
    0xdf7c,
    0xaf9b,
    0xbfba,
    0x8fd9,
    0x9ff8,
    0x6e17,
    0x7e36,
    0x4e55,
    0x5e74,
    0x2e93,
    0x3eb2,
    0x0ed1,
    0x1ef0,
  ];

  let crc = 0xffff;

  for (let i = 0; i < bytes.length; i++) {
    const c = bytes[i];
    const j = (c! ^ (crc >> 8)) & 0xff;

    crc = crcTable[j]! ^ (crc << 8);
  }

  const answer = (crc ^ 0) & 0xffff;

  const hex = answer.toString(16).padStart(4, "0").toUpperCase();

  if (invert) return hex.slice(2) + hex.slice(0, 2);

  return hex;
}


src/index.ts
import { computeCRC } from './utils/crc16'

export type Field = {
  id: string
  size: string
  value: string | Field[]
}

export type Params = {
  key: string
  amount?: string
  name: string
  city: string
}

const CRC16 = { id: '63', size: '04', value: '' }
const PAYLOAD_FORMAT_INDICATOR: Field = { id: '00', size: '02', value: '01' }
const MERCHANT_CATEGORY_CODE: Field = { id: '52', size: '04', value: '0000' }
const TRANSACTION_CURRENCY: Field = { id: '53', size: '03', value: '986' }
const COUNTRY_CODE: Field = { id: '58', size: '02', value: 'BR' }

/**
 * These fields are not mandatory, but some banks require them.
 */
const TXID: Field = { id: '05', size: '03', value: '***' }
const ADDITIONAL_DATA_FIELD_TEMPLATE = { id: '62', size: '07', value: [TXID] }

const MERCHANT_NAME = (name: string): Field => ({
  id: '59',
  size: `${name.length}`,
  value: name
})

const MERCHANT_CITY = (city: string): Field => ({
  id: '60',
  size: `${city.length}`,
  value: city
})

const MERCHANT_ACCOUNT_INFORMATION__GUI: Field = { id: '00', size: '14', value: 'br.gov.bcb.pix' }

const MERCHANT_ACCOUNT_INFORMATION__KEY = (key: string): Field => ({
  id: '01',
  size: `${key.length}`,
  value: key
})

const MERCHANT_ACCOUNT_INFORMATION = (key: string) => ({
  id: '26',
  size: '',
  value: [MERCHANT_ACCOUNT_INFORMATION__GUI, MERCHANT_ACCOUNT_INFORMATION__KEY(key)]
})

const TRANSACTION_AMOUNT = (amount: string = '00.00') => ({
  id: '54',
  size: `${amount.length}`,
  value: amount
})

function fieldToString(field: Field): string {
  if (!Array.isArray(field.value)) {
    return `${field.id}${`${field.size}`.padStart(2, '0')}${field.value}`
  }

  const value = field.value.map(fieldToString).join('')
  const size = `${value.length}`.padStart(2, '0')
  return `${field.id}${size}${value}`
}

/**
 * Generates a PIX code to be used in transferences.
 * 
 * **Some banks do not allow special characters in the PIX code.
 * Make sure you remove such carachters before calling this function**
 * @param params Transference information
 * @param params.key Pix key
 * @param params.amount Amount to be transfered
 * @param params.name Receiver name
 * @param params.city Receiver city
 * @returns 
 */
export function pix({ key, amount, name, city}: Params) {
  const fields: Field[] = [
    PAYLOAD_FORMAT_INDICATOR,
    MERCHANT_ACCOUNT_INFORMATION(key),
    MERCHANT_CATEGORY_CODE,
    TRANSACTION_CURRENCY
  ]

  if (amount) fields.push(TRANSACTION_AMOUNT(amount))

  fields.push(...[COUNTRY_CODE, MERCHANT_NAME(name), MERCHANT_CITY(city), ADDITIONAL_DATA_FIELD_TEMPLATE, CRC16])

  const payload = fields.map(fieldToString).join('')

  const crcHash = computeCRC(payload).toUpperCase()

  return `${payload}${crcHash}`
}


src/pix.ts
import { computeCRC } from "./crc16.ts";

interface Field {
  /**
   * Numeric id of the field
   */
  id: string;
  /**
   * Total field size
   */
  size: string;
  /**
   * Value of the field
   */
  value: string | Field[];
}

/**
 * Params to encode in the pix string
 *
 * **Note that some banks don't support special characters in pix codes;
 * make sure to strip those from all fields before encoding them**
 */
export interface Params {
  /**
   * Pix key
   */
  key: string;
  /**
   * Amount to be transfered
   */
  amount?: string;
  /**
   * Receiver name
   */
  name: string;
  /**
   * Receiver city
   */
  city: string;
}

const CRC16 = { id: "63", size: "04", value: "" };
const PAYLOAD_FORMAT_INDICATOR: Field = { id: "00", size: "02", value: "01" };
const MERCHANT_CATEGORY_CODE: Field = { id: "52", size: "04", value: "0000" };
const TRANSACTION_CURRENCY: Field = { id: "53", size: "03", value: "986" };
const COUNTRY_CODE: Field = { id: "58", size: "02", value: "BR" };

/**
 * These fields are not mandatory, but some banks require them.
 */
const TXID: Field = { id: "05", size: "03", value: "***" };
const ADDITIONAL_DATA_FIELD_TEMPLATE = { id: "62", size: "07", value: [TXID] };

const MERCHANT_NAME = (name: string): Field => ({
  id: "59",
  size: `${name.length}`,
  value: name,
});

const MERCHANT_CITY = (city: string): Field => ({
  id: "60",
  size: `${city.length}`,
  value: city,
});

const MERCHANT_ACCOUNT_INFORMATION__GUI: Field = {
  id: "00",
  size: "14",
  value: "br.gov.bcb.pix",
};

const MERCHANT_ACCOUNT_INFORMATION__KEY = (key: string): Field => ({
  id: "01",
  size: `${key.length}`,
  value: key,
});

const MERCHANT_ACCOUNT_INFORMATION = (key: string) => ({
  id: "26",
  size: "",
  value: [
    MERCHANT_ACCOUNT_INFORMATION__GUI,
    MERCHANT_ACCOUNT_INFORMATION__KEY(key),
  ],
});

const TRANSACTION_AMOUNT = (amount: string = "00.00") => ({
  id: "54",
  size: `${amount.length}`,
  value: amount,
});

function fieldToString(field: Field): string {
  if (!Array.isArray(field.value)) {
    return `${field.id}${`${field.size}`.padStart(2, "0")}${field.value}`;
  }

  const value = field.value.map(fieldToString).join("");
  const size = `${value.length}`.padStart(2, "0");
  return `${field.id}${size}${value}`;
}

/**
 * Generates a PIX code to be used in transferences.
 *
 * **Some banks do not allow special characters in the PIX code.
 * Make sure you remove such characters from all fields before calling this function**
 * @param params Transference information
 * @returns
 */
export function pix(params: Params) {
  const { key, amount, name, city } = params;
  const fields: Field[] = [
    PAYLOAD_FORMAT_INDICATOR,
    MERCHANT_ACCOUNT_INFORMATION(key),
    MERCHANT_CATEGORY_CODE,
    TRANSACTION_CURRENCY,
  ];

  if (amount) fields.push(TRANSACTION_AMOUNT(amount));

  fields.push(...[
    COUNTRY_CODE,
    MERCHANT_NAME(name),
    MERCHANT_CITY(city),
    ADDITIONAL_DATA_FIELD_TEMPLATE,
    CRC16,
  ]);

  const payload = fields.map(fieldToString).join("");

  const crcHash = computeCRC(payload).toUpperCase();

  return `${payload}${crcHash}`;
}
